data-science-onramp/data-cleaning/clean.py

```diff
@@ -123,69 +123,61 @@ def compute_end_udf(duration, start, end):
+        print(f'{TABLE} does not exist.')
+        raise 
 
-    # Single-parameter column transformations
+    # Single-parameter udfs
     udfs = {
-        'start_station_name': (station_name_udf, StringType()),
-        'end_station_name': (station_name_udf, StringType()),
-        'tripduration': (trip_duration_udf, IntegerType()),
-        'usertype': (user_type_udf, StringType()),
-        'gender': (gender_udf, StringType()),
-        'start_station_latitude': (angle_udf, FloatType()),
-        'start_station_longitude': (angle_udf, FloatType()),
-        'end_station_latitude': (angle_udf, FloatType()),
-        'end_station_longitude': (angle_udf, FloatType())
+        'start_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'end_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'tripduration': UserDefinedFunction(trip_duration_udf, IntegerType()),
+        'usertype': UserDefinedFunction(user_type_udf, StringType()),
+        'gender': UserDefinedFunction(gender_udf, StringType()),
+        'start_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'start_station_longitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_longitude': UserDefinedFunction(angle_udf, FloatType())
     }
 
     for name, udf in udfs.items():
-        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+        df = df.withColumn(name, udf(name))
 
-    # Multiple-parameter column transformations
+    # Multi-parameter udfs
     multi_udfs = {
         'tripduration': {
-            'udf': (compute_duration_udf, IntegerType()),
+            'udf': UserDefinedFunction(compute_duration_udf, IntegerType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         },
         'starttime': {
-            'udf': (compute_start_udf, StringType()),
+            'udf': UserDefinedFunction(compute_start_udf, StringType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         },
         'stoptime': {
-            'udf': (compute_end_udf, StringType()),
+            'udf': UserDefinedFunction(compute_end_udf, StringType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         }
     }
 
     for name, obj in multi_udfs.items():
-        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+        df = df.withColumn(name, obj['udf'](*obj['params']))
 
-    if upload:
-        upload_to_gcs(df)
-    
+    # Display sample of 100 rows
     df.sample(False, 0.001).show(n=100)
 
-if __name__ == '__main__':
-    main()
+    # Write results to GCS
+    if '--test' in sys.argv:
+        print('Data will not be uploaded to GCS')
+    else:
+        path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+        df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path, mode='overwrite')
```

Comment:
Is it considered good practice to overwrite a file instead of printing out an error that it already exists?

Reply:
Probably not, but I figured for the scope of the tutorial that it wouldn't be an issue. What do you think?

Reply:
Honestly not sure. Why do we even need to add the overwrite?

Reply:
Otherwise it throws errors if run more than once